### PR TITLE
Query: Refactoring: First part of Include/Nav unification.

### PR DESCRIPTION
**Pull request by anpete**
Replaces separate Include code path with use of Navigation rewrite pipeline via new IncludeCompiler. This first part deals with Include chains that contain no collection nav props.

- Performance: New IncludeCompiler generates optimal Include/Fixup/Tracking code. No metadata access at execution time.
- Many Include queries are simplified.
- No more duplicate joins between navs and Includes and for common prefix Include paths.

TODO:

- Uncovered #7569 and #7573 - nav rewrite issues.
- Handle collection navs by enhancing nav rewrite so that it can do Include style multi-queries.

### CODE REVIEW

File: src/Microsoft.EntityFrameworkCore/Query/IncludeSpecification.cs [hunk 1]
```diff
@@ -6,6 +6,7 @@
 using JetBrains.Annotations;
 using Microsoft.EntityFrameworkCore.Internal;
 using Microsoft.EntityFrameworkCore.Metadata;
+using Microsoft.EntityFrameworkCore.Query.ResultOperators.Internal;
```
Comment (maumar) [inline src/Microsoft.EntityFrameworkCore/Query/IncludeSpecification.cs]: remove

File: test/Microsoft.EntityFrameworkCore.SqlServer.FunctionalTests/IncludeSqlServerTest.cs [hunk 2]
```diff
@@ -155,58 +156,71 @@ public override void Include_reference_and_collection(bool useString)
         {
             base.Include_reference_and_collection(useString);
 
-            Assert.Equal(
-                @"SELECT [o].[OrderID], [o].[CustomerID], [o].[EmployeeID], [o].[OrderDate], [c].[CustomerID], [c].[Address], [c].[City], [c].[CompanyName], [c].[ContactName], [c].[ContactTitle], [c].[Country], [c].[Fax], [c].[Phone], [c].[PostalCode], [c].[Region]
+            if (!useString)
```
Comment (maumar) [inline test/Microsoft.EntityFrameworkCore.SqlServer.FunctionalTests/IncludeSqlServerTest.cs]: we should also keep baselines for the string version

File: test/Microsoft.EntityFrameworkCore.SqlServer.FunctionalTests/IncludeSqlServerTest.cs [hunk 3]
```diff
@@ -155,58 +156,71 @@ public override void Include_reference_and_collection(bool useString)
...
-LEFT JOIN [Customers] AS [c] ON [o].[CustomerID] = [c].[CustomerID]
+LEFT JOIN [Customers] AS [o.Customer] ON [o].[CustomerID] = [o.Customer].[CustomerID]
 ORDER BY [o].[OrderID]
 
 SELECT [o0].[OrderID], [o0].[ProductID], [o0].[Discount], [o0].[Quantity], [o0].[UnitPrice]
 FROM [Order Details] AS [o0]
 WHERE EXISTS (
     SELECT 1
     FROM [Orders] AS [o]
+    LEFT JOIN [Customers] AS [o.Customer] ON [o].[CustomerID] = [o.Customer].[CustomerID]
```
Comment (maumar) [inline test/Microsoft.EntityFrameworkCore.SqlServer.FunctionalTests/IncludeSqlServerTest.cs]: will this still work for required navigation?

File: src/Microsoft.EntityFrameworkCore/Query/Internal/IncludeCompiler.cs [hunk 4]
```diff
@@ -0,0 +1,417 @@
...
+                if (ReferenceEquals(querySourceReferenceExpression, _querySourceReferenceExpression))
+                {
+                    _querySourceReferenceExpression = null;
+
+                    return _includeExpression;
+                }
+
+                return querySourceReferenceExpression;
+            }
+        }
```
Comment (maumar) [inline src/Microsoft.EntityFrameworkCore/Query/Internal/IncludeCompiler.cs]: Don't we need VisitSubquery here as well?

Comment (anpete) [inline src/Microsoft.EntityFrameworkCore/Query/Internal/IncludeCompiler.cs]: No, these query sources can only originate from the outer-most query.

File: src/Microsoft.EntityFrameworkCore/Query/Internal/IncludeCompiler.cs [hunk 5]
```diff
@@ -0,0 +1,417 @@
...
+                        targetEntityExpression,
+                        relatedArrayAccessExpression));
+            }
+            else
+            {
+                blockExpressions.Add(
+                    Expression.Assign(
+                        Expression.MakeMemberAccess(
+                            targetEntityExpression,
+                            navigation.GetMemberInfo(false, true)),
```
Comment (smitpatel) [inline src/Microsoft.EntityFrameworkCore/Query/Internal/IncludeCompiler.cs]: parameter names

**Pull request merged**